dataset/src/main/scala/frameless/TypedWindow.scala

```diff
@@ -0,0 +1,156 @@
+    .orderBy(orderSpec.map(e => new Column(e.expr)):_*)
+  //TODO: frame
+
+
+  def partitionBy[U](
+    column: TypedColumn[T, U]
+  ): TypedWindow[T, A with PartitionedWindow] =
+    partitionByMany(column)
+
+  def partitionBy[U, V](
+    column1: TypedColumn[T, U],
+    column2: TypedColumn[T, V]
+  ): TypedWindow[T, A with PartitionedWindow] =
+    partitionByMany(column1, column2)
+
+  def partitionBy[U, V, W](
+    column1: TypedColumn[T, U],
+    column2: TypedColumn[T, V],
+    column3: TypedColumn[T,W]
+  ): TypedWindow[T, A with PartitionedWindow] =
+    partitionByMany(column1, column2, column3)
+
+  object partitionByMany extends ProductArgs {
+    def applyProduct[U <: HList](columns: U)
+      (implicit
+        i1: ToTraversable.Aux[U, List, TypedColumn[T, _]]
+      ): TypedWindow[T, A with PartitionedWindow] = {
+      new TypedWindow[T, A with PartitionedWindow](
+        partitionSpec = columns.toList[TypedColumn[T, _]],
+        orderSpec = orderSpec,
+        frame = frame
+      )
+    }
+  }
+
+  def orderBy[U](
+    column: SortedTypedColumn[T, U]
+  ): TypedWindow[T, A with OrderedWindow] =
+    orderByMany(column)
+
+  def orderBy[U, V](
+    column1: SortedTypedColumn[T, U],
+    column2: SortedTypedColumn[T, V]
+  ): TypedWindow[T, A with OrderedWindow] =
+    orderByMany(column1, column2)
+
+  def orderBy[U, V, W](
+    column1: SortedTypedColumn[T, U],
+    column2: SortedTypedColumn[T, V],
+    column3: SortedTypedColumn[T, W]
+  ): TypedWindow[T, A with OrderedWindow] =
+    orderByMany(column1, column2, column3)
+
+  object orderByMany extends ProductArgs {
+    def applyProduct[U <: HList, O <: HList](columns: U)
+      (implicit
+        i0: Mapper.Aux[SortedTypedColumn.defaultAscendingPoly.type, U, O],
+        i1: ToTraversable.Aux[O, List, SortedTypedColumn[T, _]]
+      ): TypedWindow[T, A with OrderedWindow] = {
+      new TypedWindow[T, A with OrderedWindow](
+        partitionSpec = partitionSpec,
+        orderSpec = i0(columns).toList[SortedTypedColumn[T, _]],
+        frame = frame
+      )
+    }
+  }
+}
+
+object TypedWindow {
+
+  def orderBy[T](
+    column: SortedTypedColumn[T, _]
+  ): TypedWindow[T, OrderedWindow] =
+    new orderByManyNew[T].apply(column) //TODO: This is some ugly syntax
+
+  def orderBy[T](
+    column1: SortedTypedColumn[T, _],
+    column2: SortedTypedColumn[T, _]
+  ): TypedWindow[T, OrderedWindow] =
+    new orderByManyNew[T].apply(column1, column2)
+
+  def orderBy[T](
+    column1: SortedTypedColumn[T, _],
+    column2: SortedTypedColumn[T, _],
+    column3: SortedTypedColumn[T, _]
+  ): TypedWindow[T, OrderedWindow] =
+    new orderByManyNew[T].apply(column1, column2, column3)
+
+  //Need different name because companion class has `orderByMany` defined as well
+  //Need a class and not object in order to define what `T` is explicitly. Otherwise it's a mess
+  //This makes for some pretty horrid syntax though.
```

Comment:
Very open to suggestions on how to make this syntax nicer. https://github.com/typelevel/frameless/pull/248/files#diff-c5eeb2825da1c4337deff8bccfefb194R93 is pretty gross in my view.